pkgs/by-name/en/encore/package.nix

```diff
@@ -0,0 +1,48 @@
+{ fetchFromGitHub, buildGoModule, callPackage, makeWrapper, lib, ... }:
+buildGoModule rec {
+  pname = "encore";
+  version = "1.39.0";
+
+  src = fetchFromGitHub {
+    name = "encore-source";
+    owner = "encoredev";
+    repo = "encore";
+    rev = "v${version}";
+    hash = "sha256-71vzo52vV4VilBnLZxIWDKIY08rfTavMZ57qgi4pip8=";
+  };
+
+  doCheck = true;
+
+  subPackages =
+    [ "cli/cmd/encore" "cli/cmd/git-remote-encore" "cli/cmd/tsbundler-encore" ];
```

Comment:
The following suggestion just 💅 feel free to disregard 🙂 
```suggestion
    map (x: "cli/cmd/" + x) [ "encore" "git-remote-encore" "tsbundler-encore" ];
```

Reply:
very nixification :v

Reply:
@afh I prefer to leave it like this for now

Reply:
Thanks for considering it, @luisnquin.

Do I understand [fmway's comment correctly](https://github.com/NixOS/nixpkgs/pull/328519#issuecomment-2240531333) that more work on this PR is necessary in order to install `tsparser-encore` or is this ready for final review?

Reply:
@afh I'm currently trying to build `tsparser-encore` too but it seems that it may take me some time

Here's my current progress

![image](https://github.com/user-attachments/assets/8b749e87-6f4b-40f6-a230-c79bb70b6961)

I have a problem trying to get the correct hashes

![image](https://github.com/user-attachments/assets/a403c891-8eea-46e9-956e-36376659474a)

nix keeps complaining about it

Reply:
Happy to try and help with this one, @luisnquin, mind updating this PR with what you have (even it is broken) or pasting a diff in a comment?

Reply:
@afh you can find it here
https://github.com/luisnquin/nixpkgs-extra/blob/main/pkgs/applications/development/tsparser-encore.nix

Reply:
Thanks, @luisnquin.

The [`src` of tsparser-encore.nix](https://github.com/luisnquin/nixpkgs-extra/blob/main/pkgs/applications/development/tsparser-encore.nix#L13-L14) suggests to me that https://github.com/encoredev/tsparser should exists, yet it 404s. What am I missing?

Reply:
Sorry it is a directory in the repository

Reply:
Not sure I understand, @luisnquin, could elaborate, please? 🙂

Reply:
First thing I found is:
```diff
- repo = "tsparser";
+ repo = "encore";
```

in https://github.com/luisnquin/nixpkgs-extra/blob/main/pkgs/applications/development/tsparser-encore.nix#L14

Reply:
@ath sorry, I was in middle of something :(

But you're right, I was in such a hurry that I made that mistake

Reply:
No worries, @luisnquin, I'm not sure how much time I have left for this today… Much success to you if you have more time in your day of this. I look forward to bringing this to  the finish line 🏁 🙂 and will check back tomorrow

Reply:
I'd like to propose the following changes contained in [encore-tsparser.patch](https://github.com/user-attachments/files/16319243/encore-tsparser.patch), @luisnquin, which:

* Rename `goEncore` to just `go` (feel free to ignore, it just made more sense to me)
* Passes encore's `version` and `src` to `tsparser.nix` to ensure consistency and avoid code duplication
* Includes `tsparser-Cargo.lock` instead of just `Cargo.lock` to be more specific to which sub-package the Cargo.lock files applies

Unfortunately with these changes I get the following build error:

```
> error: failed to select a version for the requirement `tracing = "^0.1.40"`
> candidate versions found which didn't match: 0.1.37
```

❓ Is this caused by an inconsistency upstream where
[`Cargo.toml` specifies version 0.1.40](https://github.com/encoredev/encore/blob/main/tsparser/Cargo.toml#L49) of `tracing` yet [`Cargo.lock` references version 0.1.37](https://github.com/encoredev/encore/blob/main/tsparser/Cargo.lock#L1595-L1596)?

Reply:
It is actually correct to use the root Cargo.lock file as this is a cargo workspace, so each sub project share the same lock file, see https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html

The tsparser/Cargo.lock is just a leftover from before we migrated to the workspace setup. We will remove that file

Reply:
Thanks @fredr, that's helpful and good to know.

@luisnquin unfortunately my rust and nixpkgs rust knowledge is too limited to know how `tsparser` from `encore` cargo workspace would best be packaged in nix. Have you come across something that could be helpful here?

Reply:
@afh I managed to get it working with some workarounds but it looks completely weird to me, it doesn't complain about a missing _cargo hash_

```nix
{
  rustPlatform,
  pkg-config,
  protobuf,
  encoreSrc,
  openssl,
  version,
  ...
}: let
  cargoLockFile = builtins.path {
    name = "tsparser-encore-cargo-lock";
    path = ./tsparser-Cargo.lock;
  };
in
  rustPlatform.buildRustPackage {
    pname = "tsparser-encore";
    version = version;

    src = encoreSrc;

    patchPhase = ''
      runHook prePatch

      rm -rf tsparser/Cargo.lock
      cp ${cargoLockFile} Cargo.lock

      rm -rf runtimes

      substituteInPlace Cargo.toml \
        --replace 'members = ["runtimes/core", "runtimes/js", "tsparser"]' 'members = ["tsparser"]' \

      substituteInPlace tsparser/src/bin/tsparser-encore.rs \
        --replace 'Some(err) => err.0.as_str(),' 'Some(err) => err.0.clone(),'

      substituteInPlace tsparser/src/bin/tsparser-encore.rs \
        --replace 'None => &format!("{:?}", err),' 'None => format!("{:?}", err),'

      runHook postPatch
    '';

    preBuild = ''
      export PROTOC="${protobuf}/bin/protoc"
    '';

    buildAndTestSubdir = "tsparser";

    nativeBuildInputs = [
      pkg-config
      openssl
    ];

    buildInputs = [
      protobuf
    ];

    doCheck = false;

    cargoLock = {
      lockFile = cargoLockFile;
      outputHashes = {
        "ast_node-0.9.5" = "sha256-EQGbqlwINoCgWEblUyuRcrxwg29K4tYpexhmGHPY15Q=";
        "postgres-protocol-0.6.6" = "sha256-1acUGWeNOqIteie78++U+g7ghBTsOLUzSRZ2AQlDbXA=";
      };
    };
  }
```

Reply:
Oh, I think that it is [no longer necessary when a lock file is passed](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/rust.section.md#importing-a-cargolock-file-importing-a-cargolock-file), still, this seems somewhat tricky to maintain to me

Reply:
Thanks for this, @luisnquin.

At first glance I have the following comments:

- Rename the `encoreSrc` parameter just `src`
- Use `inherit version src;` instead of assigning `version` and `encoreSrc`
- A comment on the Cargo.lock operations (`rm` and `cp`) as well as on each `substituteInPlace` would be very helpful
- Use `"${lib.getExe protobuf}"` instead of `"${protobuf}/bin/protoc"`

What will be tricky to maintain from your point of view?

Reply:
At the moment, the way how the Cargo.lock file is being configured and the "subtitute in place" statements

Reply:
Can you provide more context as to why they are necessary? Maybe these should be raised with upstream, unless they are specific to nixpkgs…

Reply:
Because there's a lifetime error in the Rust code, I'm not sure if that only happens on Nix since I wasn't able to build it with my normal shell due to a lack of time

![image](https://github.com/user-attachments/assets/33a210d1-1e4d-4633-9502-2b8d669fe480)